File: osgi-framework-bootstrap/src/test/java/net/corda/osgi/framework/OSGiBundleMock.java
package net.corda.osgi.framework;

import org.osgi.framework.Bundle;
import org.osgi.framework.BundleContext;
import org.osgi.framework.ServiceReference;
import org.osgi.framework.Version;

import java.io.File;
import java.io.InputStream;
import java.net.URL;
import java.security.cert.X509Certificate;
import java.util.Dictionary;
import java.util.Enumeration;
import java.util.Hashtable;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

final class OSGiBundleMock implements Bundle {
    private final long id;
    private final String location;

    //: Bundle

    private final AtomicInteger stateAtomic;

    OSGiBundleMock(long id, String location) {
        this.id = id;
        this.location = location;
        stateAtomic = new AtomicInteger(Bundle.INSTALLED);
    }

    /**
     * See {@link Bundle#getState}.
     *
     * @return An element of {@link Bundle#UNINSTALLED}, {@link Bundle#INSTALLED}, {@link Bundle#RESOLVED},
     *                       {@link Bundle#STARTING}, {@link Bundle#STOPPING}, {@link Bundle#ACTIVE}.
     */
    @Override
    public int getState() {
        return stateAtomic.get();
    }

    //: Comparable

    /**
     * See {@link Comparable#compareTo}
     *
     * @param other bundle to compare.
     * @return a negative integer, zero, or a positive integer as this object is
     *         less than, equal to, or greater than the specified object.
     */
    @Override
    public int compareTo(Bundle other) {
        return Long.signum(id - other.getBundleId());
    }

    public void start(int ignored) {
        start();
    }

    @Override
    public void start() {
        stateAtomic.set(Bundle.ACTIVE);
    }

    @Override
    public void stop(int options) {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public void stop() {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public void update(InputStream input) {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public void update() {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public void uninstall() {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public Dictionary<String, String> getHeaders() {
        return new Hashtable<>();
    }

    @Override
    public Dictionary<String, String> getHeaders(String ignored) {
        return getHeaders();
    }

    @Override
    public long getBundleId() {
        return id;
    }

    @Override
    public String getLocation() {
        return location;
    }

    @Override
    public ServiceReference<?>[] getRegisteredServices() {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public ServiceReference<?>[] getServicesInUse() {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public boolean hasPermission(Object permission) {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public URL getResource(String name) {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public String getSymbolicName() {
        return "mock-symbolic-name";
    }

    @Override
    public Class<?> loadClass(String name) {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    public Enumeration<URL> getResources(String name) {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public Enumeration<String> getEntryPaths(String path) {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public URL getEntry(String path) {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public long getLastModified() {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public Enumeration<URL> findEntries(String path, String filePattern, boolean recurse) {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public BundleContext getBundleContext() {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    public Map<X509Certificate, List<X509Certificate>> getSignerCertificates(int signersType) {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public Version getVersion() {
        return new Version(0, 0, 0, "mock");
    }

    @Override
    public <A> A adapt(Class<A> type) {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public File getDataFile(String filename) {
        throw new UnsupportedOperationException("Not yet implemented");
    }
}

File: osgi-framework-bootstrap/src/test/kotlin/net/corda/osgi/framework/OSGiBundleMock.kt
package net.corda.osgi.framework

import org.osgi.framework.Bundle
import org.osgi.framework.BundleContext
import org.osgi.framework.ServiceReference
import org.osgi.framework.Version
import java.io.File
import java.io.InputStream
import java.net.URL
import java.security.cert.X509Certificate
import java.util.*
import java.util.concurrent.atomic.AtomicInteger

class OSGiBundleMock(
    private val id: Long,
    private val location: String
): Bundle {

    //: Bundle

    private val stateAtomic = AtomicInteger(Bundle.INSTALLED)

    /**
     * See [Bundle.getState].
     *
     * @return An element of [Bundle.UNINSTALLED], [Bundle.INSTALLED], [Bundle.RESOLVED], [Bundle.STARTING],
     *                       [Bundle.STOPPING], [Bundle.ACTIVE].
     */
    override fun getState(): Int {
        return stateAtomic.get()
    }

    //: Comparable

    /**
     * See [Comparable.compareTo]
     *
     * @param other bundle to compare.
     * @return a negative integer, zero, or a positive integer as this object is
     *         less than, equal to, or greater than the specified object.
     */
    override fun compareTo(other: Bundle): Int {
        val thisId = this.bundleId
        val thatId = other.bundleId
        return if (thisId < thatId) -1 else if (thisId == thatId) 0 else 1
    }


    override fun start(ignored: Int) {
        start()
    }

    override fun start() {
        stateAtomic.set(Bundle.ACTIVE)
    }

    override fun stop(options: Int) {
        TODO("Not yet implemented")
    }

    override fun stop() {
        TODO("Not yet implemented")
    }

    override fun update(input: InputStream?) {
        TODO("Not yet implemented")
    }

    override fun update() {
        TODO("Not yet implemented")
    }

    override fun uninstall() {
        TODO("Not yet implemented")
    }

    override fun getHeaders(): Dictionary<String, String> {
        return Hashtable()
    }

    override fun getHeaders(ignored: String?): Dictionary<String, String> {
        return getHeaders()
    }

    override fun getBundleId(): Long {
        return id
    }

    override fun getLocation(): String {
        return location
    }

    override fun getRegisteredServices(): Array<ServiceReference<*>> {
        TODO("Not yet implemented")
    }

    override fun getServicesInUse(): Array<ServiceReference<*>> {
        TODO("Not yet implemented")
    }

    override fun hasPermission(permission: Any?): Boolean {
        TODO("Not yet implemented")
    }

    override fun getResource(name: String?): URL {
        TODO("Not yet implemented")
    }

    override fun getSymbolicName(): String {
        return "mock-symbolic-name"
    }

    override fun loadClass(name: String?): Class<*> {
        TODO("Not yet implemented")
    }

    override fun getResources(name: String?): Enumeration<URL> {
        TODO("Not yet implemented")
    }

    override fun getEntryPaths(path: String?): Enumeration<String> {
        TODO("Not yet implemented")
    }

    override fun getEntry(path: String?): URL {
        TODO("Not yet implemented")
    }

    override fun getLastModified(): Long {
        TODO("Not yet implemented")
    }

    override fun findEntries(path: String?, filePattern: String?, recurse: Boolean): Enumeration<URL> {
        TODO("Not yet implemented")
    }

    override fun getBundleContext(): BundleContext {
        TODO("Not yet implemented")
    }

    override fun getSignerCertificates(signersType: Int): MutableMap<X509Certificate, MutableList<X509Certificate>> {
        TODO("Not yet implemented")
    }

    override fun getVersion(): Version {
        return Version(0, 0, 0, "mock")
    }

    override fun <A : Any?> adapt(type: Class<A>?): A {
        TODO("Not yet implemented")
    }

    override fun getDataFile(filename: String?): File {
        TODO("Not yet implemented")
    }
}